packages/rx-effects/src/mvc.ts

```diff
@@ -72,37 +72,43 @@ export function declareViewController<
   factory: (
     deps: Dependencies,
     scope: Scope,
-  ) => Service | ((scope: Scope, ...params: Params) => Service),
+  ) => ((scope: Scope, ...params: Params) => Service) | Service,
 ): ViewControllerFactory<Service, Params>;
 
 export function declareViewController<
   Dependencies extends DependencyProps,
   Service extends AnyObject,
   Params extends unknown[],
-  Factory extends (
-    deps: Dependencies,
-    scope: Scope,
-  ) => Service | ((scope: Scope, ...params: Params) => Service),
+  Factory extends (scope: Scope, ...params: Params) => Service,
+  FactoryWithDependencies extends
+    | ((deps: Dependencies, scope: Scope) => Service)
+    | ((
+        deps: Dependencies,
+        scope: Scope,
+      ) => (scope: Scope, ...params: Params) => Service),
 >(
   tokensOrFactory: TokenProps<Dependencies> | Factory,
-  factory?: Factory,
+  factory?: FactoryWithDependencies,
 ): ViewControllerFactory<Service, Params> {
-  const factoryValue = (factory ?? tokensOrFactory) as Factory;
-
   const tokensValue = (
     factory ? tokensOrFactory : {}
```

Comment:
Hmm, it may be worth to check a type of `tokensOrFactory` here and cache `typeof` result for the 100 line...

Reply:
We need know when using factory function with scope or with dependencies.
The new commit decides some questions https://github.com/mnasyrov/rx-effects/pull/11/commits/fe601babe410a44dfd5c60cba26ddeb45c04b448